File: scripts/benchmark_host_helper.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import TypedDict


class RelativeLayout(TypedDict):
    infra_root: str
    terraform_root: str
    ansible_root: str
    infra_readme_path: str
    todo_path: str


class TargetInfo(TypedDict):
    provider: str
    os: str


class BenchmarkHostPaths(TypedDict):
    repo_root: str
    relative_layout: RelativeLayout
    recommended_stack: list[str]
    packer_stage: str
    target: TargetInfo


def benchmark_host_paths() -> BenchmarkHostPaths:
    repo_root = Path(__file__).resolve().parent.parent
    relative_layout: RelativeLayout = {
        "infra_root": "infra",
        "terraform_root": "infra/terraform",
        "ansible_root": "infra/ansible",
        "infra_readme_path": "infra/README.md",
        "todo_path": "TODO2.md",
    }
    return {
        "repo_root": str(repo_root),
        "relative_layout": relative_layout,
        "recommended_stack": ["terraform", "ansible", "apt-snapshots"],
        "packer_stage": "deferred",
        "target": {
            "provider": "hetzner",
            "os": "ubuntu-24.04",
        },
    }


def _render_human(payload: BenchmarkHostPaths) -> str:
    relative_layout = payload["relative_layout"]
    target = payload["target"]
    lines = [
        f"repo_root: {payload['repo_root']}",
        f"infra_root: {relative_layout['infra_root']}",
        f"terraform_root: {relative_layout['terraform_root']}",
        f"ansible_root: {relative_layout['ansible_root']}",
        f"infra_readme_path: {relative_layout['infra_readme_path']}",
        f"todo_path: {relative_layout['todo_path']}",
        "recommended_stack: " + ", ".join(payload["recommended_stack"]),
        f"packer_stage: {payload['packer_stage']}",
        f"target_provider: {target['provider']}",
        f"target_os: {target['os']}",
    ]
    return "\n".join(lines)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Report the canonical benchmark-host layout for this repository."
    )
    parser.add_argument("--json", action="store_true", help="Emit machine-readable JSON.")
    args = parser.parse_args()

    payload = benchmark_host_paths()
    if args.json:
        print(json.dumps(payload, indent=2, sort_keys=True))
        return
    print(_render_human(payload))


if __name__ == "__main__":
    main()

File: xtask/src/benchmark_host_helper.rs
use serde::Serialize;

use crate::repo_root;

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct RelativeLayout {
    // Keep this declaration order aligned with Python's json.dumps(sort_keys=True).
    pub ansible_root: &'static str,
    pub infra_readme_path: &'static str,
    pub infra_root: &'static str,
    pub terraform_root: &'static str,
    pub todo_path: &'static str,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct TargetInfo {
    // Keep this declaration order aligned with Python's json.dumps(sort_keys=True).
    pub os: &'static str,
    pub provider: &'static str,
}

#[derive(Debug, Clone, Eq, PartialEq, Serialize)]
pub struct BenchmarkHostPaths {
    // Keep this declaration order aligned with Python's json.dumps(sort_keys=True).
    pub packer_stage: &'static str,
    pub recommended_stack: [&'static str; 3],
    pub relative_layout: RelativeLayout,
    pub repo_root: String,
    pub target: TargetInfo,
}

pub fn benchmark_host_paths() -> BenchmarkHostPaths {
    BenchmarkHostPaths {
        packer_stage: "deferred",
        recommended_stack: ["terraform", "ansible", "apt-snapshots"],
        relative_layout: RelativeLayout {
            ansible_root: "infra/ansible",
            infra_readme_path: "infra/README.md",
            infra_root: "infra",
            terraform_root: "infra/terraform",
            todo_path: "TODO2.md",
        },
        repo_root: repo_root().display().to_string(),
        target: TargetInfo {
            os: "ubuntu-24.04",
            provider: "hetzner",
        },
    }
}

pub fn render_human(payload: &BenchmarkHostPaths) -> String {
    let layout = &payload.relative_layout;
    let target = &payload.target;

    [
        format!("repo_root: {}", payload.repo_root),
        format!("infra_root: {}", layout.infra_root),
        format!("terraform_root: {}", layout.terraform_root),
        format!("ansible_root: {}", layout.ansible_root),
        format!("infra_readme_path: {}", layout.infra_readme_path),
        format!("todo_path: {}", layout.todo_path),
        format!(
            "recommended_stack: {}",
            payload.recommended_stack.join(", ")
        ),
        format!("packer_stage: {}", payload.packer_stage),
        format!("target_provider: {}", target.provider),
        format!("target_os: {}", target.os),
    ]
    .join("\n")
}

#[cfg(test)]
mod tests {
    use super::benchmark_host_paths;
    use crate::repo_root;

    #[test]
    fn benchmark_host_helper_returns_canonical_paths() {
        let payload = benchmark_host_paths();

        assert_eq!(payload.repo_root, repo_root().display().to_string());
        assert_eq!(payload.relative_layout.infra_root, "infra");
        assert_eq!(payload.relative_layout.terraform_root, "infra/terraform");
        assert_eq!(payload.relative_layout.ansible_root, "infra/ansible");
        assert_eq!(payload.relative_layout.infra_readme_path, "infra/README.md");
        assert_eq!(payload.relative_layout.todo_path, "TODO2.md");
        assert_eq!(
            payload.recommended_stack,
            ["terraform", "ansible", "apt-snapshots"]
        );
        assert_eq!(payload.packer_stage, "deferred");
        assert_eq!(payload.target.provider, "hetzner");
        assert_eq!(payload.target.os, "ubuntu-24.04");
    }
}
